### PR TITLE
v3 WIP

### PR DESCRIPTION
**Pull request by luca-moser**

### CODE REVIEW

File: output_nft.go [hunk 1]
```diff
@@ -0,0 +1,424 @@
...
+				case uint32(FeatureBlockTag):
+				default:
+					return nil, fmt.Errorf("%w: unable to deserialize nft output, unsupported feature block type %s", ErrUnsupportedFeatureBlockType, FeatureBlockType(ty))
+				}
+				return FeatureBlockSelector(ty)
+			},
+			WriteGuard: func(seri serializer.Serializable) error {
+				switch seri.(type) {
+				case *SenderFeatureBlock:
+				case *IssuerFeatureBlock:
```
Comment (thibault-martinez) [inline output_nft.go:70]: This shouldn't be here ?

File: transaction.go [hunk 2]
```diff
@@ -162,313 +191,512 @@ func (t *Transaction) UnmarshalJSON(bytes []byte) error {
...
-			if usedSigBlockIndex != sigBlockIndex {
-				return 0, nil, fmt.Errorf("%w: target for UTXO %v uses a different signature unlock block (%d) than a previous UTXO (%d) for the same address", ErrInputSignatureUnlockBlockInvalid, utxoID, sigBlockIndex, usedSigBlockIndex)
+		outputSimpleTransfersPerIdent := make(map[string]uint64)
+		for _, output := range svCtx.WorkingSet.Tx.Essence.Outputs {
+			outDeposit := output.Deposit()
+			out += outDeposit
+
+			// accumulate simple transfers for StorageDepositReturnUnlockCondition checks
+			if basicOutput, is := output.(*BasicOutput); is {
+				if len(basicOutput.FeatureBlocks()) > 0 {
```
Comment (lzpap) [inline transaction.go:578]: This must have been forgotten here since we introduced unlock conditions instead of feature blocks.
The check should rather be:
 - The basic output only has one unlock condition, and that is Address Unlock.

File: transaction.go [hunk 3]
```diff
@@ -162,313 +191,512 @@ func (t *Transaction) UnmarshalJSON(bytes []byte) error {
...
 
-		var err error
-		deposit, err := utxo.Deposit()
-		if err != nil {
-			return 0, nil, fmt.Errorf("unable to get deposit from UTXO %v (input at index %d): %w", utxoID, i, err)
-		}
-		inputSum += deposit
+			if !unlockCondSet.HasExpirationCondition() {
+				continue
+			}
```
Comment (lzpap) [inline transaction.go:557]: Which case does this check implement?

Comment (luca-moser) [reply transaction.go]: Removed in [removes a short circuit check](https://github.com/iotaledger/iota.go/pull/295/commits/ab477d4e698da1d92633628aff1f1bfedd1fd00e)

File: .github/workflows/test_and_benchmark.yml [hunk 4]
```diff
@@ -4,7 +4,7 @@ jobs:
   test_and_benchmark:
     strategy:
       matrix:
-        go-version: [ 1.15.x ]
+        go-version: [ 1.17.x ]
```
[the remark this answers is not in the document]
Comment (muXxer) [reply .github/workflows/test_and_benchmark.yml:7]: Ah I think the wasp team waits until TinyGo is also 1.18 compatible.
Guess they are waiting for this PR to be merged https://github.com/tinygo-org/tinygo/pull/2725

[the remark this answers is not in the document]
Comment (dkegel-fastly) [reply .github/workflows/test_and_benchmark.yml:7]: Note that https://github.com/tinygo-org/tinygo/pull/2725 mostly just lets apps build with go 1.18 installed.  It also adds 'any' support.  It won't add generics, that will have to come later.

File: tpkg/util.go [hunk 5]
```diff
@@ -255,253 +268,124 @@ func RandMilestone(parents iotago.MessageIDs) (*iotago.Milestone, []byte) {
...
 
-// RandIndexation returns a random indexation payload.
-func RandIndexation(dataLength ...int) (*iotago.Indexation, []byte) {
-	const index = "寿司を作って"
-
+// RandTaggedData returns a random tagged data payload.
+func RandTaggedData(tag []byte, dataLength ...int) *iotago.TaggedData {
 	var data []byte
 	switch {
 	case len(dataLength) > 0:
```
Comment (muXxer) [inline tpkg/util.go:291]: Should we add the possibility to test against data with zero length?

File: nodeclient/indexer_client.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+	res := &IndexerResponse{}
+	if _, err := client.Do(ctx, http.MethodGet, route, nil, res); err != nil {
+		return nil, nil, err
+	}
+
+	if len(res.Items) == 0 {
+		return nil, nil, fmt.Errorf("%w for route %s", ErrIndexerNotFound, route)
+	}
+
+	outputID := res.Items.MustOutputIDs()[0]
```
Comment (muXxer) [inline nodeclient/indexer_client.go:158]: We really want to panic here?

File: nodeclient/indexer_models.go [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+	"github.com/pasztorpisti/qs"
+)
+
+// IndexerResponse is the standard successful response by the indexer.
+type IndexerResponse struct {
+	// The ledger index at which these outputs where available at.
+	LedgerIndex int `json:"ledgerIndex"`
+	// The maximum count of results that are returned by the node.
+	PageSize int `json:"pageSize"`
+	// The output IDs (transaction hash + output index) of the outputs on this address.
```
Comment (muXxer) [inline nodeclient/indexer_models.go:14]:
```suggestion
// The output IDs (transaction hash + output index) of the found outputs.
```

File: input.go [hunk 8]
```diff
@@ -59,20 +115,24 @@ func InputsUTXORefsUniqueValidator() InputsValidatorFunc {
...
 	}
 }
 
-var utxoInputRefBoundsValidator = InputsUTXORefIndexBoundsValidator()
+var inputsPredicateIndicesWithinBounds = InputsSyntacticalIndicesWithinBounds()
 
-// ValidateInputs validates the inputs by running them against the given InputsValidatorFunc.
-func ValidateInputs(inputs serializer.Serializables, funcs ...InputsValidatorFunc) error {
+// ValidateInputs validates the inputs by running them against the given InputsSyntacticalValidationFunc(s).
+func ValidateInputs(inputs Inputs, funcs ...InputsSyntacticalValidationFunc) error {
```
Comment (muXxer) [inline input.go:135]: Maybe add the Syntactical to the function name to make it more clear?
Just a suggestion, I know its also part of the comment.

File: input.go [hunk 9]
```diff
@@ -59,20 +115,24 @@ func InputsUTXORefsUniqueValidator() InputsValidatorFunc {
...
 
-// InputsUTXORefIndexBoundsValidator returns a validator which checks that the UTXO ref index is within bounds.
-func InputsUTXORefIndexBoundsValidator() InputsValidatorFunc {
-	return func(index int, input *UTXOInput) error {
-		if input.TransactionOutputIndex < RefUTXOIndexMin || input.TransactionOutputIndex > RefUTXOIndexMax {
+// InputsSyntacticalIndicesWithinBounds returns an InputsSyntacticalValidationFunc which checks that the UTXO ref index is within bounds.
+func InputsSyntacticalIndicesWithinBounds() InputsSyntacticalValidationFunc {
+	return func(index int, input Input) error {
+		ref, is := input.(IndexedUTXOReferencer)
+		if !is {
```
Comment (muXxer) [inline input.go:122]: Shouldn't we return an error here?
If that line is valid, please add a comment why this is fine.

File: address_test.go [hunk 10]
```diff
@@ -1,71 +1,34 @@
...
-	originEdAddr, originData := tpkg.RandEd25519Address()
-	tests := []struct {
-		name   string
-		source *iotago.Ed25519Address
-		target []byte
-	}{
 		{
-			"ok", originEdAddr, originData,
+			name:   "ok - NFTAddress",
+			source: tpkg.RandEd25519Address(),
```
Comment (muXxer) [inline address_test.go:25]: This should be a NFT address.

File: tagged_data.go [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+
+	"github.com/iotaledger/hive.go/serializer/v2"
+	"github.com/iotaledger/iota.go/v3/util"
+)
+
+const (
+	// TaggedPayloadTagMaxLength defines the max length of the tag within a TaggedData payload.
+	TaggedPayloadTagMaxLength = 64
+	// TaggedPayloadTagMinLength defines the min length of the tag within a TaggedData payload.
+	TaggedPayloadTagMinLength = 0
```
Comment (muXxer) [inline tagged_data.go:16]: Does the `TaggedPayloadTagMinLength` check make any sense if it zero?

File: milestone.go [hunk 12]
```diff
@@ -106,26 +112,20 @@ type (
...
-	// MilestoneInclusionMerkleProof is the inclusion merkle proof data of a milestone.
-	MilestoneInclusionMerkleProof = [MilestoneInclusionMerkleProofLength]byte
+	// MilestoneMerkleProof is the merkle root within a milestone.
+	MilestoneMerkleProof = [MilestoneMerkleProofLength]byte
 )
 
-// NewMilestone creates a new Milestone. It automatically orders the given public keys by their byte order.
-func NewMilestone(index uint32, timestamp uint64, parents MilestoneParentMessageIDs, inclMerkleProof MilestoneInclusionMerkleProof, pubKeys []MilestonePublicKey) (*Milestone, error) {
+// NewMilestone creates a new unsigned Milestone.
+func NewMilestone(index uint32, timestamp uint64, lastMs MilestoneID, parents MilestoneParentMessageIDs, pastConeMerkleProof MilestoneMerkleProof, inclMerkleProof MilestoneMerkleProof) (*Milestone, error) {
```
Comment (muXxer) [inline milestone.go:120]: The `error` return can be removed.

File: milestone.go [hunk 13]
```diff
@@ -135,64 +135,66 @@ type Milestone struct {
...
-	// The next minimum PoW score to use after NextPoWScoreMilestoneIndex is hit.
-	NextPoWScore uint32
-	// The milestone index at which the PoW score changes to NextPoWScore.
-	NextPoWScoreMilestoneIndex uint32
-	// The public keys validating the signatures of the milestone.
-	PublicKeys []MilestonePublicKey
-	// The inner payload of the milestone. Can be nil or a Receipt.
-	Receipt serializer.Serializable
+	// The merkle root of all directly/indirectly referenced messages (their IDs) which
+	// were newly confirmed by this milestone.
```
Comment (muXxer) [inline milestone.go:144]: We call that "referenced" in the hornet codebase. :)

File: milestone.go [hunk 14]
```diff
@@ -487,16 +475,16 @@ func (m *Milestone) UnmarshalJSON(bytes []byte) error {
...
-	InclusionMerkleProof       string           `json:"inclusionMerkleProof"`
-	NextPoWScore               int              `json:"nextPoWScore"`
-	NextPoWScoreMilestoneIndex int              `json:"nextPoWScoreMilestoneIndex"`
-	PublicKeys                 []string         `json:"publicKeys"`
-	Receipt                    *json.RawMessage `json:"receipt"`
-	Signatures                 []string         `json:"signatures"`
+	Type                 int                `json:"type"`
+	Index                int                `json:"index"`
+	Timestamp            int                `json:"timestamp"`
+	LastMilestone        string             `json:"lastMilestone"`
```
Comment (muXxer) [inline milestone.go:481]: Better use `lastMilestoneId`

File: address_alias.go [hunk 15]
```diff
@@ -0,0 +1,162 @@
...
+type AliasAddress [AliasAddressBytesLength]byte
+
+func (aliasAddr *AliasAddress) Clone() Address {
+	cpy := &AliasAddress{}
+	copy(cpy[:], aliasAddr[:])
+	return cpy
+}
+
+func (aliasAddr *AliasAddress) VByteCost(costStruct *RentStructure, _ VByteCostFunc) uint64 {
+	return costStruct.VBFactorData.Multiply(serializer.SmallTypeDenotationByteSize + AliasAddressBytesLength)
```
Comment (muXxer) [inline address_alias.go:54]: You could use `AliasAddressSerializedBytesSize` instead.

File: address_ed25519.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+type Ed25519Address [Ed25519AddressBytesLength]byte
+
+func (edAddr *Ed25519Address) Clone() Address {
+	cpy := &Ed25519Address{}
+	copy(cpy[:], edAddr[:])
+	return cpy
+}
+
+func (edAddr *Ed25519Address) VByteCost(costStruct *RentStructure, _ VByteCostFunc) uint64 {
+	return costStruct.VBFactorData.Multiply(serializer.SmallTypeDenotationByteSize + Ed25519AddressBytesLength)
```
Comment (muXxer) [inline address_ed25519.go:51]: You could use `Ed25519AddressSerializedBytesSize` instead.

File: address_nft.go [hunk 17]
```diff
@@ -0,0 +1,162 @@
...
+type NFTAddress [NFTAddressBytesLength]byte
+
+func (nftAddr *NFTAddress) Clone() Address {
+	cpy := &NFTAddress{}
+	copy(cpy[:], nftAddr[:])
+	return cpy
+}
+
+func (nftAddr *NFTAddress) VByteCost(costStruct *RentStructure, _ VByteCostFunc) uint64 {
+	return costStruct.VBFactorData.Multiply(serializer.SmallTypeDenotationByteSize + NFTAddressBytesLength)
```
Comment (muXxer) [inline address_nft.go:54]: You could use `NFTAddressSerializedBytesSize` instead.

File: address_alias.go [hunk 18]
```diff
@@ -0,0 +1,162 @@
...
+	}
+	copy(aliasAddr[:], data[serializer.SmallTypeDenotationByteSize:])
+	return AliasAddressSerializedBytesSize, nil
+}
+
+func (aliasAddr *AliasAddress) Size() int {
+	return AliasAddressSerializedBytesSize
+}
+
+func (aliasAddr *AliasAddress) Serialize(_ serializer.DeSerializationMode, deSeriCtx interface{}) (data []byte, err error) {
```
Comment (muXxer) [inline address_alias.go:106]: Move that function over `Size()`, then its the same order like in `address_nft.go` and `address_ed25519.go`

File: address_nft.go [hunk 19]
```diff
@@ -0,0 +1,162 @@
...
+// jsonNFTAddress defines the json representation of an NFTAddress.
+type jsonNFTAddress struct {
+	Type  int    `json:"type"`
+	NFTId string `json:"nftId"`
+}
+
+func (j *jsonNFTAddress) ToSerializable() (serializer.Serializable, error) {
+	addrBytes, err := DecodeHex(j.NFTId)
+	if err != nil {
+		return nil, fmt.Errorf("unable to decode address from JSON for NFT address: %w", err)
```
Comment (muXxer) [inline address_nft.go:154]: Sometimes it's `nft` and sometimes its `NFT` in the error messages.

File: chain.go [hunk 20]
```diff
@@ -0,0 +1,124 @@
...
+	// ChainTransitionTypeDestroy indicates that the chain is being destroyed.
+	ChainTransitionTypeDestroy
+)
+
+// StateTransitionValidationFunc is a function which given the current and next chain state,
+// validates the state transition.
+type StateTransitionValidationFunc func(current ChainConstrainedOutput, next ChainConstrainedOutput) error
+
+// IsIssuerOnOutputUnlocked checks whether the issuer in an IssuerFeatureBlock of this new ChainConstrainedOutput has been unlocked.
+// This function is a no-op if the chain is not new, or it does not contain an IssuerFeatureBlock.
```
Comment (muXxer) [inline chain.go:81]: The comment `is a no-op if the chain is not new` is misleading, because you don't check the ChainTransitionType inside this function. This is checked instead in the code where you call this function.

File: rent.go [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+
+// NonEphemeralObject is an object which can not be pruned by nodes as it
+// makes up an integral part to execute the IOTA protocol. This kind of objects are associated
+// with costs in terms of the resources they take up.
+type NonEphemeralObject interface {
+	// VByteCost returns the cost this object has in terms of taking up
+	// virtual and physical space within the data set needed to implement the IOTA protocol.
+	// The override parameter acts as an escape hatch in case the cost needs to be adjusted
+	// according to some external properties outside the NonEphemeralObject.
+	VByteCost(costStruct *RentStructure, override VByteCostFunc) uint64
```
Comment (muXxer) [inline rent.go:77]: Rename `costStruct` to `rentStruct`

File: rent.go [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+// VByteCostFunc is a function which computes the virtual byte cost of a NonEphemeralObject.
+type VByteCostFunc func(costStruct *RentStructure) uint64
```
Comment (muXxer) [inline rent.go:81]: Rename `costStruct` to `rentStruct`

Comment (muXxer) [reply rent.go:81]: `VByteCost` is already associated with the cost in IOTA tokens per VByte.
I would suggest to rename `VByteCostFunc` to `VBytesFunc` instead.

// VBytesFunc is a function which computes the virtual bytes size of a NonEphemeralObject.

File: rent.go (hunk 21 above)
Comment (muXxer) [reply rent.go:77]: `VByteCost` is already associated with the cost in IOTA tokens per VByte.
I would suggest to rename `VByteCost` to `VBytes` instead.
	
// VBytes returns the size this object has in terms of taking up
// virtual and physical space within the data set needed to implement the IOTA protocol.
// The override parameter acts as an escape hatch in case the cost needs to be adjusted
// according to some external properties outside the NonEphemeralObject.

File: rent.go [hunk 23]
```diff
@@ -0,0 +1,81 @@
...
+	minRent := vbcs.VByteCost * object.VByteCost(vbcs, nil)
+	if rent < minRent {
+		return 0, fmt.Errorf("%w: needed %d but only got %d", ErrVByteRentNotCovered, minRent, rent)
+	}
+	return minRent, nil
+}
+
+// MinStorageDeposit returns the minimum renting costs for an BasicOutput which returns
+// a StorageDepositReturnUnlockCondition amount back to the origin sender.
+func (vbcs *RentStructure) MinStorageDeposit(sender Address) uint64 {
```
Comment (muXxer) [inline rent.go:65]: Rename `MinStorageDeposit` to `ReturnOutputMinStorageDeposit` maybe? :)
Just cosmetics ofc.

File: output_alias.go [hunk 24]
```diff
@@ -0,0 +1,658 @@
...
+}
+
+// jsonAliasOutput defines the json representation of an AliasOutput.
+type jsonAliasOutput struct {
+	Type            int                `json:"type"`
+	Amount          string             `json:"amount"`
+	NativeTokens    []*json.RawMessage `json:"nativeTokens"`
+	AliasID         string             `json:"aliasId"`
+	StateIndex      int                `json:"stateIndex"`
+	StateMetadata   string             `json:"stateMetadata"`
```
Comment (alexsporn) [inline output_alias.go:607]:
```suggestion
	StateMetadata   string             `json:"stateMetadata,omitempty"`
```

File: tpkg/util.go (hunk 5 above)
Comment (luca-moser) [reply tpkg/util.go:302]: If dataLength[0] is zero / `RandTaggedData([]byte("bla"), 0)`, then this will return a `TaggedData` with a zero data field.

File: input_utxo.go [hunk 25]
```diff
@@ -18,87 +19,76 @@ const (
...
+func (u *UTXOInput) ID() OutputID {
+	var id OutputID
 	copy(id[:TransactionIDLength], u.TransactionID[:])
 	binary.LittleEndian.PutUint16(id[TransactionIDLength:], u.TransactionOutputIndex)
 	return id
 }
 
-func (u *UTXOInput) Deserialize(data []byte, deSeriMode serializer.DeSerializationMode) (int, error) {
+func (u *UTXOInput) Equals(other *UTXOInput) bool {
+	if u == nil {
```
[the remark this answers is not in the document]
Comment (luca-moser) [reply input_utxo.go:51]: Was added by @Juliusan; lets leave it like this as it might otherwise break wasp code. Debatable whether a `nil == nil`.

File: milestone.go (hunk 13 above)
Comment (luca-moser) [reply milestone.go:144]: You mean `// were newly confirmed by this milestone.` to `// were newly referenced by this milestone.`?

File: milestone.go (hunk 14 above)
Comment (luca-moser) [reply milestone.go]: Is is renamed

File: rent.go (hunk 23 above)
Comment (luca-moser) [reply rent.go]: e9e9364b598165fbb24366858ffcbc4eae91a134

File: rent.go (hunk 21 above)
Comment (luca-moser) [reply rent.go]: 3eda5b6fc04aad94422a34541f4c9fe3c36e953e

File: chain.go (hunk 20 above)
Comment (luca-moser) [reply chain.go]: 3db8417be9e8be007ba3fbdd8e95dcced39c692d

File: tagged_data.go (hunk 11 above)
Comment (luca-moser) [reply tagged_data.go]: 1c56bd813d8ec4a10c7e676d06cbed2f3c8ade0a

File: address_test.go (hunk 10 above)
Comment (luca-moser) [reply address_test.go]: a6e066c9f3503b4f804aedd8bedf059dfc542855

File: input.go [hunk 26]
```diff
@@ -36,21 +94,19 @@ func InputSelector(inputType uint32) (serializer.Serializable, error) {
...
-		var b strings.Builder
-		if _, err := b.Write(input.TransactionID[:]); err != nil {
-			return fmt.Errorf("%w: unable to write tx ID in ref validator", err)
-		}
-		if err := binary.Write(&b, binary.LittleEndian, input.TransactionOutputIndex); err != nil {
-			return fmt.Errorf("%w: unable to write UTXO index in ref validator", err)
+	return func(index int, input Input) error {
+		ref, is := input.(IndexedUTXOReferencer)
+		if !is {
+			return nil
```
[the remark this answers is not in the document]
Comment (luca-moser) [reply input.go]: a7300f79334604e22638bee8af6424642905025b

File: input.go (hunk 9 above)
Comment (luca-moser) [reply input.go:122]: 9b091b3f600ae9f24643c269f90c3f2965a40f23

File: input.go (hunk 8 above)
Comment (luca-moser) [reply input.go]: 276d8c15ffc6a9152f21386980dfc68f72ff5d0e

File: transaction_essence.go [hunk 27]
```diff
@@ -286,32 +321,23 @@ func (u *TransactionEssence) UnmarshalJSON(bytes []byte) error {
...
-	}
-
-	if len(u.Outputs) == 0 {
-		return ErrMinOutputsNotReached
-	}
-
-	if err := ValidateInputs(u.Inputs,
-		InputsUTXORefIndexBoundsValidator(),
-		InputsUTXORefsUniqueValidator(),
+func (u *TransactionEssence) syntacticallyValidate(protoParas *ProtocolParameters) error {
```
Comment (alexsporn) [inline transaction_essence.go:326]: This is missing the `Network ID must match the value of the current network.` syntactical validation comparing it to the given ProtocolParameters

File: input_utxo.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (Juliusan) [reply input_utxo.go:51]: This check safeguards from null pointer exceptions. IMO nothing is the same as nothing.
